device/cuda/src/details/sparse_ccl.cu

```diff
@@ -0,0 +1,343 @@
+
+        /*
+         * The algorithm finishes if an iteration leaves the arrays unchanged.
+         * This varible will be set if a change is made, and dictates if another
+         * loop is necessary.
+         */
+        bool gfc;
+
+        do {
+            /*
+             * Reset the end-parameter to false, so we can set it to true if we
+             * make a change to our arrays.
+             */
+            gfc = false;
+
+            /*
+             * The algorithm executes in a loop of three distinct parallel
+             * stages. In this first one, we examine adjacent cells and copy
+             * their cluster ID if it is lower than our, essentially merging
+             * the two together.
+             */
+            for (unsigned int tst = 0; tst * blockDim.x + threadIdx.x < size; ++tst) {
+                unsigned int tid = tst * blockDim.x + threadIdx.x;
+
+                for (unsigned int k = 0; k < adjc[tst]; ++k) {
+                    unsigned int j = adjv[tst][k];
+                    if (gf[tid] > gf[j]) {
+                        f[f[tid]] = gf[j];
+                        f[tid] = gf[j];
+                    }
+                }
+            }
+
+            /*
+             * Each stage in this algorithm must be preceded by a
+             * synchronization barrier!
+             */
+            __syncthreads();
+
+            /*
+             * The second stage is shortcutting, which is an optimisation that
+             * allows us to look at any shortcuts in the cluster IDs that we
+             * can merge without adjacency information.
+             */
+            for (unsigned int tid = threadIdx.x; tid < size; tid += blockDim.x) {
+                if (f[tid] > gf[tid]) {
+                    f[tid] = gf[tid];
+                }
+            }
+
+            /*
+             * Synchronize before the final stage.
+             */
+            __syncthreads();
+
+            /*
+             * Update the array for the next generation, keeping track of any
+             * changes we make.
+             */
+            for (unsigned int tid = threadIdx.x; tid < size; tid += blockDim.x) {
+                if (gf[tid] != f[f[tid]]) {
+                    gf[tid] = f[f[tid]];
+                    gfc = true;
+                }
+            }
+
+            /*
+             * Synchronize before we head on to the next iteration, or before we
+             * head to the post-processing stage.
+             */
+            __syncthreads();
+
+            /*
+             * To determine whether we need another iteration, we use block
+             * voting mechanics. Each thread checks if it has made any changes
+             * to the arrays, and votes. If any thread votes true, all threads
+             * will return a true value and go to the next iteration. Only if
+             * all threads return false will the loop exit.
+             */
+        } while (__any_sync(__activemask(), gfc));
```

Comment:
I can imagine what `__any_sync` does, but what does `__activemask()` to?

Reply:
The `__any_sync` function takes a thread mask, which is a 32-bit number where the _i_th bit indicates whether the _i_th thread of the 32-thread warp is currently executing. Remember, when a GPU thread skips a conditional block it doesn't just skip the program counter to the end, it just marks itself as inactive until the end of the block. Active mask gives you that mask. Requesting a response in the `__any_sync` from all the threads unconditionally can lead to dead-lock.